application/src/main/java/org/togetherjava/tjbot/commands/moderation/attachment/AttachmentListener.java

```diff
@@ -0,0 +1,118 @@
+package org.togetherjava.tjbot.commands.moderation.attachment;
+
+import net.dv8tion.jda.api.EmbedBuilder;
+import net.dv8tion.jda.api.MessageBuilder;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.entities.MessageEmbed;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
+import net.dv8tion.jda.api.requests.RestAction;
+import org.jetbrains.annotations.NotNull;
+import org.jooq.tools.StringUtils;
+import org.togetherjava.tjbot.commands.MessageReceiverAdapter;
+import org.togetherjava.tjbot.config.Config;
+import org.togetherjava.tjbot.moderation.ModAuditLogWriter;
+
+import java.awt.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+
+/**
+ * Listener that receives all sent messages in every channel and checks if the message has any
+ * blacklisted file attached.
+ * <p>
+ * If there was a blacklisted file attached, delete the message and send the user a dm, telling what
+ * they did wrong and also inform the mods.
+ */
+public class AttachmentListener extends MessageReceiverAdapter {
+    private final Config config;
+    private final ModAuditLogWriter modAuditLogWriter;
+
+    /**
+     * Creates a AttachmentListener to receive all message sent in any channel.
+     *
+     * @param config to find the blacklisted media attachments
+     * @param modAuditLogWriter to inform the mods about the suspicious attachment
+     */
+    public AttachmentListener(@NotNull Config config,
+            @NotNull ModAuditLogWriter modAuditLogWriter) {
+        super(Pattern.compile(".*"));
+        this.config = config;
+        this.modAuditLogWriter = modAuditLogWriter;
+    }
+
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
+        if (event.getAuthor().isBot() || event.isWebhookMessage()) {
+            return;
+        }
+        if (attachmentContainsBlacklistedFileExtension(event.getMessage())) {
+            deleteMessageAndDmUser(event);
+            warnMods(event);
```

Comment:
You should have a single queue of actions, not multiple.
```java
deleteMessage(event)
  .flatMap(any -> dmUser(event))
  .flatMap(any -> warnMods(event))
  .queue()
```

Reply:
i did not implement like this was just an example, still my opinion about this flatMap chaning is horrible to read at
 
> ```java
> deleteMessage(event)
>   .flatMap(any -> dmUser(event))
>   .flatMap(any -> warnMods(event))
>   .queue()
> ```

but again it is just an opinion 😄

Reply:
I agree, but u dont have any option here. If you dont chain it, u have 3 blocking calls on the JDA thread instead of just one.
And again, you dont want it to proceed the chain if any of them failed.

This is the proper way to do async programming with JDA 🤷